### PR TITLE
Tier roles

### LINKED ISSUE #75: Tier roles in aurora
Interns/Staff/Admin
not sure who should have what access right now but some things we should probably limit 
- [x] read only users
- [x] can't see conversation logs
- [x] can't see/create api keys

### PR DESCRIPTION
**Pull request by axton21**
# What's this PR do?

###  Adding staff and intern roles
* Having more controls over what aurora user can do on the website, giving different permissions to different roles.
* Admin - can read all info, update users information, access to keys page, can assign roles and merge users.
* Staff- can read all user info, edit users information.
* Intern - can read all user info except mobile commons messages and zendesk tickets.

# Any involved ticket?
* #75 Tier roles for aurora users.

### CODE REVIEW

**Comment by axton21** [comment]
screen shot 2015-08-05 at 11 01 33
screen shot 2015-08-05 at 11 02 09

File: app/controllers/UsersController.php [hunk 1]
```diff
@@ -149,20 +157,36 @@ public function search()
...
-    foreach($db_admins as $admin){
-      $users[] = $this->northstar->getUser('_id', $admin['_id']);
+    $employee['admin'] = User::usersWithRole('admin');
+
+    $employee['staff'] = User::usersWithRole('staff');
+
+    $employee['intern'] = User::usersWithRole('intern');
+
+    // users that tried to sign in but has no role or unauthorized
+    $employee['unassigned'] = DB::select('select * from users left join role_user on users.id = role_user.user_id where role_user.user_id is NULL');
```
Comment (angaither) [inline app/controllers/UsersController.php]: let's convert this to eloquent
`Users:join->where('role')->isNull()` or something of the likes

File: app/controllers/UsersController.php [hunk 2]
```diff
@@ -2,17 +2,18 @@
...
+use Illuminate\Support\Facades\Input;
 
 class UsersController extends \BaseController {
 
   public function __construct(NorthstarAPI $northstar) {
     $this->beforeFilter('auth');
-    $this->beforeFilter('role:admin');
+    $this->beforeFilter('roles');
+    $this->beforeFilter('intern', ['only'=>['edit', 'update', 'mobileCommonsMessages', 'zendeskTickets']]);
+    $this->beforeFilter('notAdmin', ['only' =>['destroy', 'roleCreate', 'staffIndex', 'deleteUnmergedUsers' ]]);
```
Comment (angaither) [inline app/controllers/UsersController.php]: it's a little confusing to have this be `notAdmin` can we explicitly say staff/intern roles instead?

File: app/controllers/UsersController.php [hunk 3]
```diff
@@ -149,20 +157,31 @@ public function search()
...
-    $db_admins = User::has('roles', 1)->get()->all();
-    foreach($db_admins as $admin){
-      $users[] = $this->northstar->getUser('_id', $admin['_id']);
+    $employee['admin'] = User::usersWithRole('admin');
+
+    $employee['staff'] = User::usersWithRole('staff');
+
+    $employee['intern'] = User::usersWithRole('intern');
+    // users that tried to sign in but has no role or unauthorized
+    $employee['unassigned'] = User::leftJoin('role_user', 'users.id', '=', 'role_user.user_id')->whereNull('role_user.user_id')->get();
```
Comment (axton21) [inline app/controllers/UsersController.php]: @angaither changed this to Eloquent model 😁😁

File: app/controllers/UsersController.php [hunk 4]
```diff
@@ -2,17 +2,18 @@
...
+use Illuminate\Support\Facades\Input;
 
 class UsersController extends \BaseController {
 
   public function __construct(NorthstarAPI $northstar) {
     $this->beforeFilter('auth');
-    $this->beforeFilter('role:admin');
+    $this->beforeFilter('roles');
+    $this->beforeFilter('internLimits', ['except'=>['index', 'show', 'search']]);
+    $this->beforeFilter('adminPrivileges', ['only' =>['destroy', 'roleCreate', 'staffIndex', 'deleteUnmergedUsers' ]]);
```
Comment (axton21) [inline app/controllers/UsersController.php]: @angaither Renamed these 2 filters 😁😁

File: app/views/users/zendesk-tickets.blade.php [hunk 5]
```diff
@@ -18,7 +18,7 @@
 							  		{{ isset($ticket['status']) ? ('<dt>Status: </dt><dd>' . ucfirst($ticket['status'])) : "" }}
 							  		<dt>Ticket ID: </dt><dd>{{ link_to("https://dosomethingorg1.zendesk.com/agent/tickets/" . $ticket['id'], $ticket['id']) }}</dd>
 							  		{{ isset($ticket['orgainzation_id']) ? ('<dt>Organization ID: </dt><dd>' . $ticket['organization_id']) : "" }}
-							  		{{ isset($ticket['group_id']) ? ('<dt>Group ID: </dt><dd>' . link_to("https://dosomethingorg1.zendesk.com/groups" . $ticket['group_id'], $ticket['group_id'])) : "" }}
+							  		{{ isset($ticket['group_id']) ? ('<dt>Group ID: </dt><dd>' . link_to("https://dosomethingorg1.zendesk.com/groups/" . $ticket['group_id'], $ticket['group_id'])) : "" }}
```
Comment (axton21) [inline app/views/users/zendesk-tickets.blade.php]: `/` is important or it will never work.. 😂

**Comment by angaither** [comment]
:+1:

**Pull request merged**